python/python/lance/dataset.py

```diff
@@ -628,6 +628,12 @@ def alter_columns(self, *alterations: Iterable[Dict[str, Any]]):
                 nullability is not changed. Only non-nullable columns can be changed
                 to nullable. Currently, you cannot change a nullable column to
                 non-nullable.
+            - "data_type": pyarrow.DataType, optional
+                The new data type to cast the column to. If not specified, the column
+                data type is not changed.
```

Comment:
From the user perspective this is a little vague.  Though I think we are using arrow's cast method under the hood?

What casts are valid?

Presumably casting to a wider type is fine.
Do we support casting to narrower types?
Do we support casting to unrelated types (e.g. integer to string)?
Can we cast from one struct type to a narrower struct type?
Can we add fields to a struct type?
Is this a safe cast or an unsafe cast?  Or, perhaps more accurately, what mutations or errors can happen during a cast?

Reply:
These are all good questions. I'm going to add some tests for these.

I think we should probably not do "safe" casts. That is we shouldn't produce a bunch of nulls if it fails; we should error instead.

I also left a note for my documentation PR that I should answer these questions in the docs: https://github.com/lancedb/lance/pull/1911#issuecomment-1934640134

Reply:
We cannot cast struct. But we can cast unrelated types. I'm not sure how keen I am to advertise it though, TBH.

Reply:
Not advertising for now is ok but eventually we should be clear both in documentation and in raising errors for things we don't want to support (even if the underlying arrow could support it)

Reply:
Maybe I'll just make it so it limits to upcasts/downcasts for now.

Reply:
Added additional restrictions on types of casts allowed.